README.md
```
____   _________________________   ________ ___________
\   \ /   /\_   _____/\______   \ /  _____/ \_   _____/
 \   Y   /  |    __)_  |       _//   \  ___  |    __)_ 
  \     /   |        \ |    |   \\    \_\  \ |        \ 2018 VERGE
   \___/   /_______  / |____|_  / \______  //_______  /
                   \/         \/         \/         \/ 
```
# A Python library for the VERGE Client
It is a set of Python libraries that allows easy access to the
VERGE peer-to-peer cryptocurrency client API.


Installation instructions
===========================

verge-python uses setuptools for the install script. There are no dependencies apart from Python itself.

::

  $ python setup.py build
  $ python setup.py install
  

Pypi / Cheeseshop
==================

It is possible to install the package through Pypi (cheeseshop), see http://pypi.python.org/pypi?:action=display&name=verge-python
::
 $ pip install verge-python
 # if not working, try
 $ pip install --pre verge-python

Connection to verge-qt
=========================

If you want to connect to verge-qt, add server=1 in your VERGE.conf
::

 rpcuser=vergerpcuser
 rpcpassword=A RANDOM GENERATED PASSWORD
 server=1

TODO
======
These things still have to be added:

- SSL support (including certificate verification) for managing remote verge daemons.

verge-python is a fork of bitcoin-python : https://github.com/laanwj/bitcoin-python




Changelog
2014-03-02 17:55:05 -0600 Naveen Garg 

	* add importprivkey method

2014-02-26 15:34:39 -0800 Peter Harrington 

	* Update config.py

2014-01-28 19:58:26 +0100 Jean-Christophe Saad-Dupuy 

	* updated changelog

2014-01-28 19:26:06 +0100 Jean-Christophe Saad-Dupuy 

	* Version bump (tag: v0.1.2)

2014-01-28 11:31:13 -0600 benfbat 

	* Update config.py

2014-01-27 23:08:33 +0100 Jean-Christophe Saad-Dupuy 

	* Added support for loading verge.conf under windows

2014-01-26 23:00:57 +0100 jcsaaddupuy 

	* Update README.rst

2014-01-26 23:06:44 +0100 Jean-Christophe Saad-Dupuy 

	* markdown

2014-01-26 23:05:31 +0100 Jean-Christophe Saad-Dupuy 

	* updated verge-qt connection instructions

2014-01-22 02:26:02 +0100 Jean-Christophe Saad-Dupuy 

	* fixed markdown

2014-01-22 02:24:44 +0100 Jean-Christophe Saad-Dupuy 

	* Added instruction for connection with verge-t

2014-01-21 18:51:36 +0100 Jean-Christophe Saad-Dupuy 

	* Updated documentation link

2014-01-21 18:41:36 +0100 Jean-Christophe Saad-Dupuy 

	* Updated changelog (tag: v0.1.1)

2014-01-21 18:41:09 +0100 Jean-Christophe Saad-Dupuy 

	* updated version in documentation

2014-01-21 09:55:38 +0100 jcsaaddupuy 

	* Update README.rst

2014-01-20 21:17:42 +0100 Jean-Christophe Saad-Dupuy 

	* Removed pip instructions

2014-01-20 20:52:44 +0100 Jean-Christophe Saad-Dupuy 

	* bump version

2014-01-20 20:52:18 +0100 Jean-Christophe Saad-Dupuy 

	* Removed uggly print

2014-01-20 20:43:03 +0100 jcsaaddupuy 

	* Update README.rst

2014-01-20 20:35:00 +0100 Jean-Christophe Saad-Dupuy 

	* Updated tests

2014-01-20 20:31:26 +0100 Jean-Christophe Saad-Dupuy 

	* Fixed markdown typo

2014-01-20 20:27:42 +0100 Jean-Christophe Saad-Dupuy 

	* version reset (dogedev)

2014-01-20 20:26:08 +0100 Jean-Christophe Saad-Dupuy 

	* Updated documentation

2014-01-20 19:58:38 +0100 Jean-Christophe Saad-Dupuy 

	* changed default rpc port

2014-01-20 19:41:43 +0100 Jean-Christophe Saad-Dupuy 

	* Migration bitcoin -> verge

2014-01-20 19:12:31 +0100 Jean-Christophe Saad-Dupuy 

	* Changed bitcoin -> verge

2013-12-26 10:33:05 +0100 Wladimir J. van der Laan 

	* fix documentation for sendtoaddress

2013-12-01 01:05:30 -0500 jon 

	* Changed iterkeys() to keys() to make the code python3 compatible. The method keys() exists in both python 2 and 3 but keysiter() does not exist in python 3 since keys() by defaults returns an iterator.

2013-10-29 12:02:27 -0200 Leandro Boscariol 

	* Made sure the code will raise InsufficientFunds or stop

2013-10-29 11:14:15 -0200 Leandro Boscariol 

	* Work around to raise InsufficientFunds exception when needed

2013-09-27 15:49:55 +0200 Nicolas Kuttler 

	* Add signmessage and verifymessage support

2013-09-27 14:29:31 +0200 Nicolas Kuttler 

	* Add --noremote option

2013-09-27 13:08:17 +0200 Nicolas Kuttler 

	* Make the test runs configurable

2013-09-11 18:16:59 +0200 mUniKeS 

	* Add gettxout function

2013-09-02 19:46:04 +0200 Wladimir J. van der Laan 

	* connect_to_local: check cfg for None

2013-05-30 17:59:03 +0200 Oliver Gasser 

	* Add dumpprivkey() function.

2013-05-30 16:48:03 +0200 Oliver Gasser 

	* Use items() instead of iteritems().

2013-04-25 08:27:27 +0300 Aviad Reich 

	* pep8 styling conventions

2013-04-24 22:39:24 -0600 Jeff Cook 

	* Raise a TransportException on HTTP 403.

2013-04-24 22:39:06 -0600 Jeff Cook 

	* Define TransportException exception class.

2013-04-22 14:05:03 +0400 Nikolay Belikov 

	* + Added exception wrapper into AuthServiceProxy

2013-04-21 08:56:40 -0600 Jeff Cook 

	* Inform user when 403 Forbidden is sent by bitcoind.

2013-04-21 13:07:59 +0300 Aviad Reich 

	* add tests for connect_to_remote. SSL is Faulty

2013-04-17 12:49:35 +0400 Nikolay Belikov 

	* ! Minor code style fixes

2013-04-16 18:46:46 +0400 Nikolay Belikov 

	* Major overhaul of AuthServiceProxy

2013-04-14 21:35:38 -0500 Jason Kölker 

	* Allow getbalance default account with minconf=0

2013-03-30 16:32:59 +0100 Wladimir J. van der Laan 

	* bump version number to 0.3 (tag: v0.3)

2013-03-30 16:13:06 +0100 Aviad Reich 

	* add keypoolrefill method

2013-03-20 00:33:59 +0200 Aviad Reich 

	* raw transaction support added

2013-03-21 14:41:33 +0200 Aviad Reich 

	* cherry pick to fix dont_raise so it only blocks the correct exceptions to return a boolean (origin/wallet_locking)

2013-03-29 10:15:09 +0300 Aviad Reich 

	* ignore dist and python_egg

2013-03-23 19:08:46 +0000 Aviad Reich 

	* deprecated getblocknumber - use getblockcount instead

2013-03-19 23:01:18 +0200 Aviad Reich 

	* add wallet locking/unlocking support

2013-03-19 14:42:26 +0200 Aviad Reich 

	* in listtransactions, iterate all accounts if account is None (origin/listtransactions)

2013-03-19 11:16:38 +0200 Aviad Reich 

	* listaccounts can return a dictionary from account names to their balance if the param as_dict is True, update tests. (origin/listaccounts)

2013-03-19 09:44:20 +0200 Aviad Reich 

	* add tests for listunspent and getmininginfo

2013-03-18 08:18:23 +0200 Aviad Reich 

	* refactor __repr__

2013-03-17 09:43:44 +0200 Aviad Reich 

	* flake8 style conventions fixes

2013-03-18 00:13:42 +0200 Aviad Reich 

	* add listunspent method

2013-03-17 23:46:44 +0200 Aviad Reich 

	* add support for getmininginfo

2013-03-17 16:37:09 +0200 Aviad Reich 

	* Issue #15: add from parameter to listtransactions

2013-03-15 22:14:10 +0100 David Barton 

	* Fixed usage.rst broken links

2013-03-14 09:26:23 +0200 Aviad Reich 

	* connect_to_local(): use default testnet port, when the config file has testnet=x

2013-02-10 17:35:14 +0100 Adrian 

	* Add getrawtransaction method

2013-01-16 03:28:11 +0100 Thomas Steen Rasmussen 

	* fix bug when rpc reply error is null in json

2012-12-12 00:14:34 +0100 Benoît HERVIER 

	* There isn't always an error key in resp dict

2012-12-12 00:12:40 +0100 Benoît HERVIER 

	* There isn't always an error key in resp dict

2012-12-12 00:10:41 +0100 Benoît HERVIER 

	* Allow use of https


src/vergerpc/exceptions.py
# Copyright (c) 2010 Witchspace <witchspace81@gmail.com>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
"""
Exception definitions.
"""


class VERGEException(Exception):
    """
    Base class for exceptions received from VERGE server.

    - *code* -- Error code from ``verged``.
    """
    # Standard JSON-RPC 2.0 errors
    INVALID_REQUEST  = -32600,
    METHOD_NOT_FOUND = -32601,
    INVALID_PARAMS   = -32602,
    INTERNAL_ERROR   = -32603,
    PARSE_ERROR      = -32700,

    # General application defined errors
    MISC_ERROR                  = -1  # std::exception thrown in command handling
    FORBIDDEN_BY_SAFE_MODE      = -2  # Server is in safe mode, and command is not allowed in safe mode
    TYPE_ERROR                  = -3  # Unexpected type was passed as parameter
    INVALID_ADDRESS_OR_KEY      = -5  # Invalid address or key
    OUT_OF_MEMORY               = -7  # Ran out of memory during operation
    INVALID_PARAMETER           = -8  # Invalid, missing or duplicate parameter
    DATABASE_ERROR              = -20 # Database error
    DESERIALIZATION_ERROR       = -22 # Error parsing or validating structure in raw format

    # P2P client errors
    CLIENT_NOT_CONNECTED        = -9  # VERGE is not connected
    CLIENT_IN_INITIAL_DOWNLOAD  = -10 # Still downloading initial blocks

    # Wallet errors
    WALLET_ERROR                = -4  # Unspecified problem with wallet (key not found etc.)
    WALLET_INSUFFICIENT_FUNDS   = -6  # Not enough funds in wallet or account
    WALLET_INVALID_ACCOUNT_NAME = -11 # Invalid account name
    WALLET_KEYPOOL_RAN_OUT      = -12 # Keypool ran out, call keypoolrefill first
    WALLET_UNLOCK_NEEDED        = -13 # Enter the wallet passphrase with walletpassphrase first
    WALLET_PASSPHRASE_INCORRECT = -14 # The wallet passphrase entered was incorrect
    WALLET_WRONG_ENC_STATE      = -15 # Command given in wrong wallet encryption state (encrypting an encrypted wallet etc.)
    WALLET_ENCRYPTION_FAILED    = -16 # Failed to encrypt the wallet
    WALLET_ALREADY_UNLOCKED     = -17 # Wallet is already unlocked

    def __init__(self, error):
        Exception.__init__(self, error['message'])
        self.code = error['code']


class TransportException(Exception):
    """
    Class to define transport-level failures.
    """
    def __init__(self, msg, code=None, protocol=None, raw_detail=None):
        self.msg = msg
        self.code = code
        self.protocol = protocol
        self.raw_detail = raw_detail
        self.s = """
        Transport-level failure: {msg}
        Code: {code}
        Protocol: {protocol}
        """.format(msg=msg, code=code, protocol=protocol)

    def __str__(self):
        return self.s


##### General application defined errors
class SafeMode(VERGEException):
    """
    Operation denied in safe mode (run ``verged`` with ``-disablesafemode``).
    """


class JSONTypeError(VERGEException):
    """
    Unexpected type was passed as parameter
    """
InvalidAmount = JSONTypeError  # Backwards compatibility


class InvalidAddressOrKey(VERGEException):
    """
    Invalid address or key.
    """
InvalidTransactionID = InvalidAddressOrKey  # Backwards compatibility


class OutOfMemory(VERGEException):
    """
    Out of memory during operation.
    """


class InvalidParameter(VERGEException):
    """
    Invalid parameter provided to RPC call.
    """


##### Client errors
class ClientException(VERGEException):
    """
    P2P network error.
    This exception is never raised but functions as a superclass
    for other P2P client exceptions.
    """


class NotConnected(ClientException):
    """
    Not connected to any peers.
    """


class DownloadingBlocks(ClientException):
    """
    Client is still downloading blocks.
    """


##### Wallet errors
class WalletError(VERGEException):
    """
    Unspecified problem with wallet (key not found etc.)
    """
SendError = WalletError  # Backwards compatibility


class InsufficientFunds(WalletError):
    """
    Insufficient funds to complete transaction in wallet or account
    """


class InvalidAccountName(WalletError):
    """
    Invalid account name
    """


class KeypoolRanOut(WalletError):
    """
    Keypool ran out, call keypoolrefill first
    """


class WalletUnlockNeeded(WalletError):
    """
    Enter the wallet passphrase with walletpassphrase first
    """


class WalletPassphraseIncorrect(WalletError):
    """
    The wallet passphrase entered was incorrect
    """


class WalletWrongEncState(WalletError):
    """
    Command given in wrong wallet encryption state (encrypting an encrypted wallet etc.)
    """


class WalletEncryptionFailed(WalletError):
    """
    Failed to encrypt the wallet
    """


class WalletAlreadyUnlocked(WalletError):
    """
    Wallet is already unlocked
    """


# For convenience, we define more specific exception classes
# for the more common errors.
_exception_map = {
    VERGEException.FORBIDDEN_BY_SAFE_MODE: SafeMode,
    VERGEException.TYPE_ERROR: JSONTypeError,
    VERGEException.WALLET_ERROR: WalletError,
    VERGEException.INVALID_ADDRESS_OR_KEY: InvalidAddressOrKey,
    VERGEException.WALLET_INSUFFICIENT_FUNDS: InsufficientFunds,
    VERGEException.OUT_OF_MEMORY: OutOfMemory,
    VERGEException.INVALID_PARAMETER: InvalidParameter,
    VERGEException.CLIENT_NOT_CONNECTED: NotConnected,
    VERGEException.CLIENT_IN_INITIAL_DOWNLOAD: DownloadingBlocks,
    VERGEException.WALLET_INSUFFICIENT_FUNDS: InsufficientFunds,
    VERGEException.WALLET_INVALID_ACCOUNT_NAME: InvalidAccountName,
    VERGEException.WALLET_KEYPOOL_RAN_OUT: KeypoolRanOut,
    VERGEException.WALLET_UNLOCK_NEEDED: WalletUnlockNeeded,
    VERGEException.WALLET_PASSPHRASE_INCORRECT: WalletPassphraseIncorrect,
    VERGEException.WALLET_WRONG_ENC_STATE: WalletWrongEncState,
    VERGEException.WALLET_ENCRYPTION_FAILED: WalletEncryptionFailed,
    VERGEException.WALLET_ALREADY_UNLOCKED: WalletAlreadyUnlocked,
}


def wrap_exception(error):
    """
    Convert a JSON error object to a more specific VERGE exception.
    """
    # work around to temporarily fix https://github.com/bitcoin/bitcoin/issues/3007
    if error['code'] == VERGEException.WALLET_ERROR and error['message'] == u'Insufficient funds':
        error['code'] = VERGEException.WALLET_INSUFFICIENT_FUNDS
    return _exception_map.get(error['code'], VERGEException)(error)
